src/common/hist_util.h

```diff
@@ -196,23 +196,135 @@ size_t DeviceSketch(int device,
  */
+
+
+enum BinBounds
+{
+  POWER_OF_TWO_8,
+  POWER_OF_TWO_16,
+  POWER_OF_TWO_32,
+};
+
+uint32_t getValueFromUint8(void *t, size_t i);
+/*{
+  std::cout << "\nuint8_t\n";
+  return 0;
+  //return ((uint8_t*)t)[i];
+}*/
+uint32_t getValueFromUint16(void* t, size_t i);
+/*{
+  std::cout << "\nuint32_t\n";
+  return 0;//((uint16_t*)t)[i];
+}
+*/
+uint32_t getValueFromUint32(void* t, size_t i);
+/*{
+  std::cout << "\nuint32_t\n";
+  return 0;//((uint32_t*)t)[i];
+}*/
+
+
+typedef uint32_t (*Func)(void*, size_t);
+
+struct Index
+{
+  Index(): binBound_(POWER_OF_TWO_8), p_(1), data_(nullptr), disp_(nullptr)
+  {
+    setBinBound(binBound_);
+  }
+  ~Index()
+  {
+    if(data_)
+    {
+      char* delete_ptr = static_cast<char*>(data_);
+      delete [] delete_ptr;
+    }
+    if(disp_)
+      delete [] disp_;
+    data_ = nullptr;
+    disp_ = nullptr;
+  }
+  Index(const Index& i) = delete;
+  Index& operator=(Index i) = delete;
+  Index(Index&& i) = delete;
+  Index& operator=(Index&& i) = delete;
+  uint32_t operator[](size_t i) const
+  {
+    return func_(data_, i) + disp_[i%p_];
+  }
+  void setBinBound(BinBounds binBound)
+  {
+    binBound_ = binBound;
+    switch(binBound)
+    {
+      case POWER_OF_TWO_8:
+        func_ = &getValueFromUint8;
+        break;
+      case POWER_OF_TWO_16:
+        func_ = &getValueFromUint16;
+        break;
+      case POWER_OF_TWO_32:
+        func_ = &getValueFromUint32;
+        break;
+      default:
+        func_ = &getValueFromUint8;
+    }
+  }
+  void setDispSize(size_t p)
+  {
+    p_ = p;
+  }
+  BinBounds getBinBound() const
+  {
+    return binBound_;
+  }
+  template<typename T>
+  T* data() const
+  {
+    return static_cast<T*>(data_);
+  }
+  uint32_t* disp() const
+  {
+    return disp_;
+  }
+  void resize(size_t nBytesData, size_t nDisps)
+  {
+    data_ = aligned_alloc(64, nBytesData);
+    disp_ = (uint32_t*)aligned_alloc(64, sizeof(uint32_t)*nDisps);
+  }
+  private:
+    void* data_;
+    uint32_t* disp_;
+    size_t p_;
+    BinBounds binBound_;
+    Func func_;
+};
+
+
 struct GHistIndexMatrix {
   /*! \brief row pointer to rows by element position */
   std::vector<size_t> row_ptr;
+
   /*! \brief The index data */
-  std::vector<uint32_t> index;
+  //std::vector<uint32_t> index;
+
+  Index index;
   /*! \brief hit count of each index */
   std::vector<size_t> hit_count;
   /*! \brief The corresponding cuts */
   HistogramCuts cut;
   // Create a global histogram matrix, given cut
   void Init(DMatrix* p_fmat, int max_num_bins);
+
+  template<typename T>
+  void setIndexData(T* const index_data, size_t batch_threads, const SparsePage& batch, size_t rbegin, uint32_t* disps, size_t nbins);
+
   // get i-th row
-  inline GHistIndexRow operator[](size_t i) const {
-    return {&index[0] + row_ptr[i],
-            static_cast<GHistIndexRow::index_type>(
-                row_ptr[i + 1] - row_ptr[i])};
-  }
+  //inline GHistIndexRow operator[](size_t i) const {
```

Comment:
Please, remove this and other commented code

Reply:
fixed